registry/storage/garbagecollect.go

```diff
@@ -47,6 +49,17 @@ func MarkAndSweep(ctx context.Context, storageDriver driver.StorageDriver, regis
 		}
 
 		err = manifestEnumerator.Enumerate(ctx, func(dgst digest.Digest) error {
+			if manifestswithouttags {
+				tags, errr := repository.Tags(ctx).Lookup(ctx, distribution.Descriptor{Digest: dgst})
+				if errr != nil {
+					return fmt.Errorf("failed to retrieve tags for digest %v: %v", dgst, errr)
+				}
+				if len(tags) == 0 {
+					emit("manifest has no tags %s@%s", repoName, dgst)
+					manifestSet[dgst] = repoName
```

Comment:
Because you are deleting the manifests after a shallow reference test (ie tags), I think you can get away with immediate deletion. Double check in paths.go.